### PR TITLE
[src/software/sfm] Merge SfM engines in a common SfM binary #1872

### PR DESCRIPTION
**Pull request by pmoulon**
Proposal to merge the SfM engine into a single binary.
It could help to bring clarity and a safe space to add new engine.

### CODE REVIEW

**Comment by pmoulon** [comment]
It would be most likely a good idea to factorize some code for some parameter we ask often (sfm_data, matches... ) to be sure we have the same parameter name in the command line...

File: docs/sphinx/rst/software/SfM/IncrementalSfM.rst [hunk 1]
```diff
@@ -80,7 +76,7 @@ Arguments description:
       - 4: Pinhole radial 3 + tangential 2
       - 5: Pinhole fisheye
 
-  - **[-f|--refineIntrinsics]**
+  - **[-f|--sIntrinsic_refinement_options]**
```
Comment (rjanvier) [inline docs/sphinx/rst/software/SfM/IncrementalSfM.rst:79]: --refine_intrinsic_config

File: src/software/SfM/SfM_GlobalPipeline.py.in [hunk 2]
```diff
@@ -71,19 +71,9 @@ if not os.path.exists(reconstruction_dir):
     os.mkdir(reconstruction_dir)
 
 print ("6. Do Global reconstruction")
-pRecons = subprocess.Popen( [os.path.join(OPENMVG_SFM_BIN, "openMVG_main_GlobalSfM"),  "-i", matches_dir+"/sfm_data.json", "-m", matches_dir, "-o", reconstruction_dir] )
+pRecons = subprocess.Popen( [os.path.join(OPENMVG_SFM_BIN, "openMVG_main_SfM"), "--engine", "GLOBAL", "--input_file", matches_dir+"/sfm_data.json", "--match_file", matches_dir+"/matches.e.bin", "--output_dir", reconstruction_dir] )
```
Comment (rjanvier) [inline src/software/SfM/SfM_GlobalPipeline.py.in:74]: Maybe using Unix directory separators in command line arguments is not portable here

File: src/software/SfM/SfM_SequentialPipeline.py.in [hunk 3]
```diff
@@ -71,19 +71,9 @@ if not os.path.exists(reconstruction_dir):
     os.mkdir(reconstruction_dir)
 
 print ("6. Do Sequential/Incremental reconstruction")
-pRecons = subprocess.Popen( [os.path.join(OPENMVG_SFM_BIN, "openMVG_main_IncrementalSfM"),  "-i", matches_dir+"/sfm_data.json", "-m", matches_dir, "-o", reconstruction_dir] )
+pRecons = subprocess.Popen( [os.path.join(OPENMVG_SFM_BIN, "openMVG_main_SfM"), "--engine", "INCREMENTAL", "--input_file", matches_dir+"/sfm_data.json", "--match_dir", matches_dir, "--output_dir", reconstruction_dir] )
```
Comment (rjanvier) [inline src/software/SfM/SfM_SequentialPipeline.py.in:74]: the previous comment also hold here.

File: src/software/SfM/main_SfM.cpp [hunk 4]
```diff
@@ -0,0 +1,589 @@
...
+    {"STELLAR", ESfMSceneInitializer::INITIALIZE_STELLAR},
+  };
+  const auto it  = string_to_enum_mapping.find(str);
+  if (it == string_to_enum_mapping.end())
+    return false;
+  scene_initializer = it->second;
+  return true;
+}
+
+/// From 2 given image file-names, find the two corresponding index in the View list
```
Comment (rjanvier) [inline src/software/SfM/main_SfM.cpp:97]: filenames

File: src/software/SfM/main_SfM.cpp [hunk 5]
```diff
@@ -0,0 +1,589 @@
...
+  // Global SfM
+  cmd.add( make_option('r', rotation_averaging_method, "rotationAveraging") );
+  cmd.add( make_option('t', translation_averaging_method, "translationAveraging") );
+
+  try {
+    if (argc == 1) throw std::string("Invalid parameter.");
+    cmd.process(argc, argv);
+  } catch (const std::string& s) {
+
+     cmd.add( make_option('i', filename_sfm_data, "input_file") );
```
Comment (rjanvier) [inline src/software/SfM/main_SfM.cpp:200]: indent ;)

File: src/software/SfM/main_SfM.cpp [hunk 6]
```diff
@@ -0,0 +1,589 @@
...
+
+
+
+    OPENMVG_LOG_INFO << "Usage: " << argv[0] << '\n'
+    << "[Required]\n"
+    << "[-i|--input_file] path to a SfM_Data scene\n"
+    << "[-m|--match_dir] path to the matches that corresponds to the provided SfM_Data scene\n"
+    << "[-o|--output_dir] path where the output data will be stored\n"
+    << "[-s|--sfm_engine] Type of SfM Engine to use for the reconstruction\n"
+      << "\t INCREMENTAL   : add image sequentially to a 2view seed\n"
```
Comment (rjanvier) [inline src/software/SfM/main_SfM.cpp:232]: space : "2 view seed"

File: src/software/SfM/main_SfM.cpp [hunk 7]
```diff
@@ -0,0 +1,589 @@
...
+    << "\t\t" << static_cast<int>(ETriangulationMethod::LINFINITY_ANGULAR) << ": LINFINITY_ANGULAR\n"
+    << "\t\t" << static_cast<int>(ETriangulationMethod::INVERSE_DEPTH_WEIGHTED_MIDPOINT) << ": INVERSE_DEPTH_WEIGHTED_MIDPOINT\n"
+    << "\t[--resection_method] resection/pose estimation method (default=" << resection_method << "):\n"
+    << "\t\t" << static_cast<int>(resection::SolverType::DLT_6POINTS) << ": DIRECT_LINEAR_TRANSFORM 6Points | does not use intrinsic data\n"
+    << "\t\t" << static_cast<int>(resection::SolverType::P3P_KE_CVPR17) << ": P3P_KE_CVPR17\n"
+    << "\t\t" << static_cast<int>(resection::SolverType::P3P_KNEIP_CVPR11) << ": P3P_KNEIP_CVPR11\n"
+    << "\t\t" << static_cast<int>(resection::SolverType::P3P_NORDBERG_ECCV18) << ": P3P_NORDBERG_ECCV18\n"
+    << "\t\t" << static_cast<int>(resection::SolverType::UP2P_KUKELOVA_ACCV10)  << ": UP2P_KUKELOVA_ACCV10 | 2Points | upright camera\n"
+    << "\n\n"
+    << "[Global]\n"
```
Comment (rjanvier) [inline src/software/SfM/main_SfM.cpp:305]: [GLOBAL]

File: src/software/SfM/tutorial_demo.py.in [hunk 8]
```diff
@@ -63,7 +63,7 @@ pFiltering.wait()
 
 reconstruction_dir = os.path.join(output_dir,"reconstruction_sequential")
 print ("3. Do Incremental/Sequential reconstruction") #set manually the initial pair to avoid the prompt question
-pRecons = subprocess.Popen( [os.path.join(OPENMVG_SFM_BIN, "openMVG_main_IncrementalSfM"),  "-i", matches_dir+"/sfm_data.json", "-m", matches_dir, "-o", reconstruction_dir] )
+pRecons = subprocess.Popen( [os.path.join(OPENMVG_SFM_BIN, "openMVG_main_SfM"), "--engine", "INCREMENTAL", "--input_file", matches_dir+"/sfm_data.json", "--match_dir", matches_dir, "--output_dir", reconstruction_dir] )
```
Comment (rjanvier) [inline src/software/SfM/tutorial_demo.py.in:66]: portable separator in the command line argument. it should be easy with Pathlib

File: src/software/SfM/tutorial_demo.py.in [hunk 9]
```diff
@@ -83,7 +83,7 @@ pFiltering.wait()
 
 reconstruction_dir = os.path.join(output_dir,"reconstruction_global")
 print ("3. Do Global reconstruction")
-pRecons = subprocess.Popen( [os.path.join(OPENMVG_SFM_BIN, "openMVG_main_GlobalSfM"),  "-i", matches_dir+"/sfm_data.json", "-m", matches_dir, "-o", reconstruction_dir] )
+pRecons = subprocess.Popen( [os.path.join(OPENMVG_SFM_BIN, "openMVG_main_SfM"), "--engine", "GLOBAL", "--input_file", matches_dir+"/sfm_data.json", "--match_file", matches_dir+"/matches.e.bin", "--output_dir", reconstruction_dir] )
```
Comment (rjanvier) [inline src/software/SfM/tutorial_demo.py.in:86]: same as previous comment

**Review by rjanvier — COMMENTED** [review]
Hey @pmoulon, thanks :+1:  ! I made a quick review and I found few minor things. Some of them were here before this PR... As a side note, I suggest we use Pathlib  everywhere in python script for path resolution if it's not a too hard requirement (Python>=3.4).

File: src/software/SfM/SfM_GlobalPipeline.py.in (hunk 2 above)
Comment (pmoulon) [reply src/software/SfM/SfM_GlobalPipeline.py.in:74]: Tested on Windows and it was ok

**Comment by pmoulon** [comment]
Thank you @rjanvier, I also saw I made a mistake on the doc for refineX in main_SfM. Fixed now ;-)

**Comment by pmoulon** [comment]
Thank you @rjanvier @rperrot for your comments. I think we are good to merge now.

**Pull request merged**